pubsublite/ps/message.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package ps
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/protobuf/proto"
+
+	tspb "github.com/golang/protobuf/ptypes/timestamp"
+	pb "google.golang.org/genproto/googleapis/cloud/pubsublite/v1"
+)
+
+const eventTimestampAttributeKey = "x-goog-pubsublite-event-time-timestamp-proto"
+
+var (
+	errInvalidMessage = errors.New("pubsublite: invalid received message")
+)
+
+// Encodes a timestamp in a way that it will be interpreted as an event time if
+// published on a message with an attribute named eventTimestampAttributeKey.
+func encodeEventTimestamp(eventTime *tspb.Timestamp) (string, error) {
+	bytes, err := proto.Marshal(eventTime)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(bytes), nil
+}
+
+// Decodes a timestamp encoded with encodeEventTimestamp.
+func decodeEventTimestamp(value string) (*tspb.Timestamp, error) {
+	bytes, err := base64.StdEncoding.DecodeString(value)
+	if err != nil {
+		return nil, err
+	}
+	eventTime := &tspb.Timestamp{}
+	if err := proto.Unmarshal(bytes, eventTime); err != nil {
+		return nil, err
+	}
+	return eventTime, nil
+}
+
+// extractOrderingKey extracts the ordering key from the message for routing
+// during publishing. It is the default KeyExtractorFunc implementation.
+func extractOrderingKey(msg *pubsub.Message) []byte {
+	if len(msg.OrderingKey) == 0 {
+		return nil
+	}
+	return []byte(msg.OrderingKey)
+}
+
+// transformPublishedMessage is the default PublishMessageTransformerFunc
+// implementation.
+func transformPublishedMessage(from *pubsub.Message, extractKey KeyExtractorFunc) (*pb.PubSubMessage, error) {
+	msgpb := &pb.PubSubMessage{
+		Data: from.Data,
+		Key:  extractKey(from),
+	}
+
+	if len(from.Attributes) > 0 {
+		msgpb.Attributes = make(map[string]*pb.AttributeValues)
+		for key, value := range from.Attributes {
+			if key == eventTimestampAttributeKey {
+				eventpb, err := decodeEventTimestamp(value)
+				if err != nil {
+					return nil, err
+				}
+				msgpb.EventTime = eventpb
+			} else {
+				msgpb.Attributes[key] = &pb.AttributeValues{Values: [][]byte{[]byte(value)}}
+			}
+		}
+	}
+	return msgpb, nil
+}
+
+// transformReceivedMessage is the default ReceiveMessageTransformerFunc
+// implementation.
+func transformReceivedMessage(from *pb.SequencedMessage, to *pubsub.Message) error {
```

Comment:
Is there a reason why transformPublishedMessage returns a tuple of (message, error) and transformReceivedMessage takes the output message as a pointer, and just returns an error? Can we normalize this?

Reply:
Good point, normalized to output pointers.

transformReceivedMessage must be provided an output message because users will not be able to create a new `pubsub.Message` from scratch (we will need to create a private ack handler implementation).